### PR TITLE
Revise append.covariates to fix #264

### LINKED ISSUE #264: append.covariates() working as expected?
Debugging something else I found that `append.covariates()` (in file `db/R/query.trait.data.R`) doesn't seem to work right. The argument `covariates.data` is a data frame (at leas when called from within `query.trait.data()`, where it was generated by `query.covariates()`), so the index for the for loop 

```for(i in seq(covariates.data)){ ``` 

is just 

```1:ncol(covariates.data) ``` 

or `1:3` in my case (maybe always?). As a result, only three rows in `data` actually have covariates appended, even when many more exist.

I think many examples would show this. I noticed it while running query.traits for leaf respiration for the temperate.Early_Hardwood PFT.

### PR DESCRIPTION
**Pull request by ryankelly-uiuc**
Per issue #264, I noticed months ago that this function was doing something weird. Finally got back to it, figured out what was intended, and fixed so that's what happens.

I had nothing to do with the original version of these functions, so it would be good if someone more familiar (e.g. @dlebauer) checked out my changes carefully. But I think the intent was pretty straightforward, and my revised code is a lot simpler than before. 

The way inputs to `append.covariates()` work is still convoluted. Note that `arrhenius.scaling.traits()` calls the function with `covariates.data` equal to a list of data frames, whereas `filter.sunleaf.traits` calls it with a single data frame passed in via `...`, which is then converted to a list because `covariates.data` is missing. Both of these uses work fine, but note that if `covariates.data` was simply a data frame the function would fail. I thought about fixing this but thought maybe there was a good reason it was done this way at some time in the past. (To me, the whole thing would be simpler if `covariates.data` was just a data frame, set up beforehand with covariates in descending order of priority. Then `append.covariates()` could use `!duplicated` to quickly pull out just the highest-priority covariate available for any trait. No need for a list at all, which would also simplify `arrhenius.scaling.traits()`, the one place where a list input is currently used.)

### CODE REVIEW

**Comment by mdietze** [comment]
looks good to me

File: db/R/query.trait.data.R [hunk 1]
```diff
@@ -121,21 +121,21 @@ query.yields <- function(trait = 'yield', spstr, extra.columns='', con=NULL, ...
 ##' @param covariates.data one or more tables of covariate data, ordered by the precedence
 ##' they will assume in the event a trait has covariates across multiple tables.
 ##' All tables must contain an 'id' and 'level' column, at minimum.
+##'
+##' @author <unknown>, Ryan Kelly
```
Comment (dlebauer) [inline db/R/query.trait.data.R]: `<unknown>` == Carl Davidson

**Comment by ryankelly-uiuc** [comment]
Couldn't help it. I revised again following my comment above. Checked and results are identical, but the implementation is a lot less convoluted. 

And noted the Unknown Author.

**Comment by ryankelly-uiuc** [comment]
Wait, one more thing...

**Comment by ryankelly-uiuc** [comment]
OK. My way required something a little convoluted too, to sort the covariates by order of priority. Still a little safer than the old way though, I think.

**Comment by dlebauer** [comment]
This looks okay, but I would expect something like this to work (values scaled to 5C > values scaled to 4C). If you are up for learning something new ... try adding this to [db/tests/testthat/test.derive.traits.R](https://github.com/PecanProject/pecan/blob/master/db/tests/testthat/test.derive.traits.R). 

```r
library(PEcAn.DB)
library(RPostgreSQL)

con <- db.open(list(driver = "PostgreSQL", user = "bety", dbname = "bety", password = "bety"))
# variable_ids: 
# Vcmax = 4; airT = 86, leafT = 81
test.traits <- db.query("select * from traits where variable_id = 4 and id in (select trait_id from covariates where variable_id in (81, 86)) and mean > 0 limit 50;", con = con)
test.cov <- query.covariates(test.traits$id, con = con)
test.traits.5C <- arrhenius.scaling.traits(data = test.traits, covariates = test.cov, temp.covariates = 'leafT', new.temp = 5)
test.traits.4C <- arrhenius.scaling.traits(data = test.traits, covariates = test.cov, temp.covariates = 'leafT', new.temp = 4)
expect_true(all(test.traits.5C$mean > test.traits.4C$mean))
```

**Comment by ryankelly-uiuc** [comment]
Cool. It does seem to work though, if you change the > to ≥:
```expect_true(all(test.traits.5C$mean >= test.traits.4C$mean))```

This is needed because some traits don't have a covariate, so don't get scaled. 

In checking this I discovered that the result of `arrhenius.scaling.traits()` has rows in different order than the input (i.e. in your example, `all(test.traits$id==test.traits.5C$id)` is `FALSE`). Think that's a problem? I'll try to look and see, and/or could modify the function to reset the row order to be same as input.

I'll check out adding this to tests...

**Comment by dlebauer** [comment]
I don't think I'd order is important since the output data frame replaces
the input data frame.

should we drop values that need a covariate but don't have one?

**Comment by ryankelly-uiuc** [comment]
> should we drop values that need a covariate but don't have one?

Well, it certainly has the potential to screw results up pretty bad if we don't. But if we can assume that usually a missing covariate was probably the default (e.g., 25 degC currently hard-coded in arrhenius scaling fn) then we might not have big errors and get to keep a lot more data.

**Comment by dlebauer** [comment]
If we make that assumption, shouldn't the test I wrote pass using ">" instead of ">=" ?

**Comment by ryankelly-uiuc** [comment]
I mean, in the actual PEcAn code, everywhere `arrhenius.scaling.traits()` is called, the new temperature is left at the default (25 degC). If we ever use the function to scale to a different temp (as in your example), then we would want to either delete traits with no covariates (as you suggested), or explicitly set their temperature to an assumed 25 degC (or whatever) before scaling.

**Comment by ryankelly-uiuc** [comment]
Woohoo tests!

Anyway, I do think the *safest* thing to do Re: missing covariates is to throw out any trait that should have a covariate but doesn't (as you suggested above). I just thought that this must have come up before, and that having the code set up to assume 25*C for missing temp covariates must have therefore been a conscious decision.

**Comment by dlebauer** [comment]
> assume 25*C for missing temp covariates must have therefore been a conscious decision.

I think it was, in general we embed assumptions into PEcAn rather than making them in the database (that way the assumption may be wrong, but at least the original data isn't). But even if we assume the trait is 25, it should be able to scale to any other temperature. 

> Woohoo tests!

Nice work - you are a natural!

**Comment by ryankelly-uiuc** [comment]
>I think it was, in general we embed assumptions into PEcAn rather than making them in the database (that way the assumption may be wrong, but at least the original data isn't). But even if we assume the trait is 25, it should be able to scale to any other temperature. 

OK, then I will change how the missing covariate case is handled. Previously, a trait with a missing covariate was just left unchanged, which works fine since throughout PEcAn the function is only ever used to scale to 25 degC. I'll change it so that a missing covariate is assumed to be 25 degC, and then those traits get scaled like all the rest.

**Comment by dlebauer** [comment]
> since throughout PEcAn the function is only ever used to scale to 25 degC

except in [write.configs.ED2](https://github.com/PecanProject/pecan/blob/master/models/ed/R/write.configs.ed.R#L49)

```r
    trait.samples[['root_respiration_rate']] <- arrhenius.scaling(rrr2, old.temp = 25, 
                                                                  new.temp = 15)
```

**Comment by ryankelly-uiuc** [comment]
but that doesn't use `arrhenius.scaling.traits()`, which is where the sneaky assumption comes in. (sneaky, because it isn't documented or commented). `arrhenius.scaling()` has a default `new.temp` (again, 25 degC), but will throw an error if `old.temp` isn't passed.

**Comment by ryankelly-uiuc** [comment]
I think this pull request is good to go but slipped through the cracks. Unless I'm missing something.

**Comment by dlebauer** [comment]
> I think this pull request is good to go but slipped through the cracks. Unless I'm missing something.

Yes slipped through the cracks. It is a good idea to check back!

**Pull request merged**